### PR TITLE
Add coverage for gRPC reflection services

### PR DESCRIPTION
**Pull request by xtroitgeo**
### Summary

Coverage for gRPC reflection services

Please select the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Dependency update
- [ ] Refactoring
- [ ] Backport
- [x] New scenario (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
- [ ] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [x] Methods and classes used in PR scenarios are meaningful
- [ ] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

### CODE REVIEW

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/GrpcResource.java [hunk 1]
```diff
@@ -20,6 +26,47 @@ public class GrpcResource {
...
+        GrpcReflectionResponse response = new GrpcReflectionResponse(invoke(request));
+        return response.getServiceList();
+    }
+
+    @GET
+    @Path("reflection/service/description")
+    @Produces(MediaType.TEXT_PLAIN)
+    public String serviceMethods() {
+        ServerReflectionRequest request = ServerReflectionRequest.newBuilder()
+                .setHost("localhost").setFileByFilename("helloworld.proto").build();
```
Comment (michalvavrik) [inline http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/GrpcResource.java:58]: `helloworld.proto` and "" for `setListServices`. What about if you expose one endpoint that accepts nullable filename and returns `GrpcReflectionResponse`?

**Review by michalvavrik — CHANGES_REQUESTED** [review]
Hey @xtroitgeo ,

thanks for the PR. I'll do proper review in the next round (when you address my comments). As for the native failures, that's upstream bug, please open issue in Quarkus project. Until it is fixed, you can do something like

```Java
import io.quarkus.runtime.annotations.RegisterForReflection;

@RegisterForReflection(classNames = "io.grpc.reflection.v1.FileDescriptorResponse")
public class ReflectionConfiguration { }
```

please read more here https://quarkus.io/guides/writing-native-applications-tips#registerForReflection.

Squash commits please.

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/GrpcResource.java [hunk 2]
```diff
@@ -20,6 +28,50 @@ public class GrpcResource {
...
+                .setHost("localhost").setFileByFilename(protoFilename + ".proto").build();
+
+        ServerReflectionResponse response = invoke(request);
+        return response.getFileDescriptorResponse().toByteArray();
+    }
+
+    private ServerReflectionResponse invoke(ServerReflectionRequest request) {
+        return reflection.serverReflectionInfo(Multi.createFrom().item(request))
+                .collect().first()
+                .await().indefinitely();
```
Comment (michalvavrik) [inline http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/GrpcResource.java:63]: Wonderful opportunity to use reactive type :-) Please get rid of `.await().indefinitely()` and return `Multi<Byte[]>`. I don't see a reason to block inside the RESTEasy Reactive resource method.

**Review by michalvavrik — CHANGES_REQUESTED** [review]
Hey @xtroitgeo ,

fine job. I'll requested one comment and one tiny little change, but apart of that, I think you are done.

Thank you,

Michal

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java [hunk 3]
```diff
@@ -107,6 +115,104 @@ public void testGrpc() {
...
+                serviceMethods = HelloWorldProto.getDescriptor().getServices()
+                        .get(service.getIndex()).getMethods();
+                break;
+            }
+        }
+
+        // Check if exists service "Greeter"
+        assertNotNull(serviceMethods);
+
+        /// Check if exists sayHello method
```
Comment (michalvavrik) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java:210]: `///` => `//`

File: http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/GrpcResource.java [hunk 4]
```diff
@@ -20,6 +28,50 @@ public class GrpcResource {
     @GrpcClient("hello")
     Greeter client;
 
+    @Inject
+    @GrpcClient("reflection-service")
+    MutinyServerReflectionGrpc.MutinyServerReflectionStub reflection;
+
+    @GET
+    @Path("reflection/service/info")
```
Comment (michalvavrik) [inline http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/GrpcResource.java:36]: Both `@Path`s you added do not start with a slash. That's absolutely fine (though personally I'm used to leading slash), but we need to stick to one convention inside the file (or ideally module).

There is already endpoint `hello` inside this resource whose path starts with the slash. Please see https://github.com/quarkus-qe/quarkus-test-suite/blob/main/http/http-advanced-reactive/src/main/java/io/quarkus/ts/http/advanced/reactive/GrpcResource.java#L24

I suggest `/reflection/service/info` and `reflection/descriptor/{profFilename}` (or change the `hello` path).

Thanks

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java [hunk 5]
```diff
@@ -107,6 +115,105 @@ public void testGrpc() {
...
     }
 
+    @Test
+    @DisplayName("GRPC reflection test - service count")
+    public void testReflection_serviceCount() {
+        GrpcReflectionResponse response = getApp().given().when().get("/api/grpc/reflection/service/info")
+                .then().statusCode(SC_OK).extract().response()
+                .jsonPath().getObject(".", GrpcReflectionResponse.class);
+
+        assertEquals(response.getServiceCount(), 3);
```
Comment (michalvavrik) [inline http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java:125]: @xtroitgeo in order to get correct validation error message, the format must be `assertEquals(expected, actual)`, please check out formal parameter names of the `assertEquals`.  When tests are failing, it gets confusing. Please correct it.

**Review by michalvavrik — CHANGES_REQUESTED** [review]

**Comment by michalvavrik** [comment]
I canceled CI in order to not wait resources (as you will kindly address my comment).

File: http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java (hunk 5 above)
Comment (xtroitgeo) [reply http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/BaseHttpAdvancedReactiveIT.java]: Fixed

**Review by michalvavrik — APPROVED** [review]
Thanks for your contribution.

**Pull request merged**